src/test/compile-fail/lifetime-inference-give-expl-lifetime-param.rs

```diff
@@ -39,8 +39,7 @@ struct Cat<'x, T> { cat: &'x isize, t: T }
 struct Dog<'y> { dog: &'y isize }
 
 fn cat2<'x, 'y>(x: Cat<'x, Dog<'y>>) -> &'x isize {
-//~^ HELP: consider using an explicit lifetime parameter as shown: fn cat2<'x>(x: Cat<'x, Dog<'x>>) -> &'x isize
-    x.t.dog //~ ERROR: cannot infer
+    x.t.dog //~ ERROR E0312
```

Comment:
@eddyb ah, yeah, it has to do with the specifics of when those messages trigger. We are generating fewer variables and edges, and this results in us being able to give a message that the heuristics think is preferable (E0312) to the more generic "cannot infer" message. In that case, we just run with that.